energymeters/integrationpluginenergymeters.cpp

```diff
@@ -0,0 +1,394 @@
+            connect(meter, &EnergyMeter::consumedEnergyReceived, this, &IntegrationPluginEnergyMeters::onConsumedEnergyReceived);
+
+            m_energyMeters.insert(info->thing(), meter);
+            info->finish(Thing::ThingErrorNoError);
+        });
+        meter->getEnergyConsumed();
+        return;
+    } else {
+        Q_ASSERT_X(false, "setupThing", QString("Unhandled thingClassId: %1").arg(thing->thingClassId().toString()).toUtf8());
+    }
+}
+
+void IntegrationPluginEnergyMeters::postSetupThing(Thing *thing)
+{
+    qCDebug(dcEnergyMeters) << "Post setup thing" << thing->name();
+    if (m_connectionStateTypeIds.contains(thing->thingClassId())) {
+        thing->setStateValue(m_connectionStateTypeIds.value(thing->thingClassId()), true);
+
+        if (m_energyMeters.contains(thing)) {
+            startUpdateCycle(m_energyMeters.value(thing));
+        }
+    }
+
+    if (!m_reconnectTimer) {
+        m_reconnectTimer = hardwareManager()->pluginTimerManager()->registerTimer(5);
+        connect(m_reconnectTimer, &PluginTimer::timeout, this, [this] {
+            foreach (Thing *thing, myThings()) {
+                if (m_connectionStateTypeIds.contains(thing->thingClassId())) {
+                    if (!thing->stateValue(m_connectionStateTypeIds.value(thing->thingClassId())).toBool()) {
+                        EnergyMeter *meter = m_energyMeters.value(thing);
+                        if (!meter)
+                            continue;
+
+                        startUpdateCycle(meter);
+                    }
+                }
+            }
+        });
+    }
+}
+
+void IntegrationPluginEnergyMeters::thingRemoved(Thing *thing)
+{
+    qCDebug(dcEnergyMeters()) << "Thing removed" << thing->name();
+
+    if (m_energyMeters.contains(thing)) {
+        EnergyMeter *meter = m_energyMeters.take(thing);
+        m_updateCycleInProgress.remove(meter);
+        meter->deleteLater();
+    }
+
+    if (myThings().isEmpty() && m_reconnectTimer) {
+        qCDebug(dcEnergyMeters()) << "Stopping reconnect timer";
+        hardwareManager()->pluginTimerManager()->unregisterTimer(m_reconnectTimer);
+        m_reconnectTimer = nullptr;
+    }
+}
+
+void IntegrationPluginEnergyMeters::startUpdateCycle(EnergyMeter *meter)
+{
+    if (m_updateCycleInProgress.contains(meter)) {
+        if (m_updateCycleInProgress.value(meter)) {
+            return;
+        }
+    }
+    m_updateCycleInProgress.insert(meter, true);
+    meter->getVoltageL1();
+}
+
+void IntegrationPluginEnergyMeters::updateCycleFinished(EnergyMeter *meter)
+{
+    m_updateCycleInProgress.insert(meter, false);
+
+    int updateInterval = configValue(energyMetersPluginUpdateIntervalParamTypeId).toInt();
+    QTimer::singleShot(updateInterval, meter, [this, meter] {
+        startUpdateCycle(meter); // restart update cycle
+    });
+}
+
+void IntegrationPluginEnergyMeters::onConnectionStateChanged(bool status)
+{
+    EnergyMeter *meter = static_cast<EnergyMeter *>(sender());
```

Comment:
qobject_cast<EnergyMeter *> is a bit faster :-)

Reply:
done